smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/provider/UserTuneProvider.java

```diff
@@ -46,74 +45,73 @@
+                        switch (name) {
+                        case "artist":
+                            builder.setArtist(parser.nextText());
+                            break;
+                        case "length":
+                            builder.setLength(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "rating":
+                            builder.setRating(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "source":
+                            builder.setSource(parser.nextText());
+                            break;
+                        case "title":
+                            builder.setTitle(parser.nextText());
+                            break;
+                        case "track":
+                            builder.setTrack(Integer.parseInt(parser.nextText()));
+                            break;
+                        case "uri":
+                            URI uri = new URI(parser.nextText());
+                            builder.setUri(uri);
+                            break;
+                        case "tune":
+                            if (UserTuneElement.ELEMENT.equals(name) && (tag == END_TAG)) {
+                                return builder.build();
                             }
                         }
-                    } catch (IllegalArgumentException e) {
-                        throw new XmlPullParserException("Unknown tune value: " + name + " encountered.");
-                    } catch (URISyntaxException e) {
-                        throw new IllegalArgumentException("Invalid URI");
+                        tag = parser.next();
+                        name = parser.getName();
+                        if (isNameInvalid(name)) {
+                            throw new XmlPullParserException("Invalid elements inside <tune/>");
+                        }
                     }
+                } catch (IllegalArgumentException e) {
+                    throw new XmlPullParserException("Unknown tune value: " + name + " encountered.");
+                } catch (URISyntaxException e) {
+                    throw new IllegalArgumentException("Invalid URI");
+                }
 
-                case END_TAG:
-                    if (UserTuneElement.ELEMENT.equals(name)) {
-                        return builder.build();
+            case END_TAG:
+                if (UserTuneElement.ELEMENT.equals(name)) {
+                    return builder.build();
                 }
             }
         }
     }
 
     private boolean isNameInvalid(String name) {
-        if (name.equals("artist") || name.equals("length") || name.equals("rating") || name.equals("source") || name.equals("title") || name.equals("track") || name.equals("uri") || name.equals("tune")) {
+        if (name.equals("artist") || name.equals("length") || name.equals("rating") || name.equals("source")
```

Comment:
I'd try something like `if (Arrays.asList("artist", "length", ...).contains(name))`

Reply:
Yes, this looks much bettew :)